src/main/java/com/example/textprocessingtool/Controller.java

```diff
@@ -74,21 +99,197 @@ public void handleSearch() {
             String result = dataManagement.replaceText(pattern, inputTextArea.getText(), replacement);
             resultTextFlow.getChildren().add(new Text(result));
+            inputTextArea.setText(result);
+            wordCountLabel.setText(String.valueOf(dataManagement.countWords(inputTextArea.getText())));
         } catch (RuntimeException e) {
             showAlert("Error", e.getMessage());
         }
     }
 
+    /**
+     * Handle add.
+     * add an item to the list
+     */
+    @FXML
+    public void handleAdd() {
+        String[] inputs = validateAndGetInputs();
+
+        if (inputs == null) {
+            return;
+        }
+
+        String name = inputs[0];
+        String value = inputs[1];
+
+        DataEntry newEntry = new DataEntry(name, value);
+        if (!dataEntries.contains(newEntry)) {
+            dataEntries.add(newEntry);
+            clearFields();
+        } else {
+            showAlert("Duplicate Entry", "An entry with this name already exists");
+        }
+    }
+
+    /**
+     * Handle update.
+     * Update a selected item in the L
+     */
+    @FXML
+    public void handleUpdate() {
+       try {
+           DataEntry selected = dataTable.getSelectionModel().getSelectedItem();
+           if (selected == null) {
+               showAlert("No Selection", "Please select an entry to update");
+               return;
+           }
+
+           String[] inputs = validateAndGetInputs();
+           if (inputs == null) {
+               return;
+           }
+
+           String name = inputs[0];
+           String value = inputs[1];
+
+           selected.setName(name);
+           selected.setValue(value);
+           dataTable.refresh();
+           clearFields();
+       } catch (RuntimeException e) {
+           showAlert("Error", e.getMessage());
+       }
+    }
+
+    /**
+     * Handle delete.
+     */
+    @FXML
+    public void handleDelete() {
+        DataEntry selected = dataTable.getSelectionModel().getSelectedItem();
+        if (selected == null) {
+            showAlert("No Selection", "Please select an entry to delete");
+            return;
+        }
+
+        dataEntries.remove(selected);
+        clearFields();
+    }
+
+    private void clearFields() {
+        nameField.clear();
+        valueField.clear();
+    }
+
     private void showAlert(String title, String content) {
         Alert alert = new Alert(Alert.AlertType.INFORMATION);
         alert.setTitle(title);
         alert.setContentText(content);
         alert.showAndWait();
     }
 
+    /**
+     * Use regular expression to validate that the name contains only letters and spaces
+     * @param name the name to validate
+     * @return true if the name is valid, false otherwise
+     */
+    private boolean validateName(String name) {
+        if (name == null) {
+            return true;
+        }
+        name = name.trim();
+
+        if (name.isEmpty()) {
+            return true;
+        }
```

Comment:
is there a reason why you have two `if` blocks to check for `null` and `empty` , both `returns true`. You can combine it into one `if` block

eg.

```java
if(n == null || n.isEmpty()) return true;
```